tests-mbed_hal-sleep fix

maciejbocianski opened the pull request:
### Description

- fix for deepsleep_lpticker_test (from tests-mbed_hal-sleep)
- enable tests-mbed_hal-sleep test

### Pull request type


    [x] Fix
    [ ] Refactor
    [ ] New target
    [ ] Feature
    [ ] Breaking change

maciejbocianski committed:
HAL sleep test fix

Prevents ticker upper layer initialization during test execution

maciejbocianski committed:
enable HAL sleep test

0xc0170 commented on TESTS/mbed_hal/sleep/main.cpp:
```diff
@@ -105,6 +105,11 @@ void sleep_usticker_test()
 
     const ticker_irq_handler_type us_ticker_irq_handler_org = set_us_ticker_irq_handler(us_ticker_isr);
 
+    // call ticker_read_us to initialize ticker upper layer
```
looking at the lines above, there's if #0 - should also be removed

0xc0170 replied:
To remove this, we need to have fix for `MBED_ALL_STATS_ENABLED` 

@maciejbocianski Please provide details

maciejbocianski replied:
When `MBED_ALL_STATS_ENABLED` is enable then `sleep_manager_sleep_auto` function calls some extra logging code, what makes return from `sleep` takes more time than expected

**To enable `sleep_usticker_test` test**
We could disable this flag globally for all test, or create mechanism to selectively mark tests which should be running without extra `log/stats` flags

bulislaw replied:
Could we just unset the flag instead?

maciejbocianski replied:
impossible to do, this flag is set globally for all tests

Jenkins code:
```mbed_test_build_cmd=(mbed test --compile -m ${target} -t ${toolchain} -DMBED_HEAP_STATS_ENABLED=1 -DMBED_STACK_STATS_ENABLED=1 -DMBED_TRAP_ERRORS_ENABLED=1 -DMBED_ALL_STATS_ENABLED)```

0xc0170 replied:
@studavekar Please review

0xc0170 commented:
@studavekar @OPpuolitaival @SenRamakri Please review https://github.com/ARMmbed/mbed-os/pull/7036#discussion_r191345496

Having all stats on might cause some time increase. In this case, as we expect a target to be running back from sleep in a specific period of time, adding extra stats adds delay and causing our tests to be failing.

I provide an example: a target from sleep should be running lets say <10us. If we enable stats, and they affect this functionality (adding some computation to the sleep functionality), this increases the time to be awake from 10us to higher numbers - not for all targets.  This is a grey area (I dont think we got documented how stats affects functionality - shall we ?). If I know that lets  say CPU stats can increase sleep times, then  this should be stated in the documentation. How shall we fix this type of test?

OPpuolitaival commented:
@0xc0170 I am not enough familiar with this that I can review

cmonr reviewed (dismissed):
Blocking until internal conversation has been resolved.

cmonr commented:
After internal discussion, the CI flag will remain, and a mechanism will be added to bypass the delay causes by printing stats. This PR will need to be rebased once https://github.com/ARMmbed/mbed-os/pull/7063 is in.

maciejbocianski commented:
In my opinion this PR should be merged first.
It contains fix for sleep test, that #7063 enables without fixing it

cmonr commented:
@maciejbocianski Making sure I understand correctly. I thought that #7063 was going to add a mechanism to bypass the additional time delay, and then this PR would use that mechanism to run the test with proper timings.

Is this not correct?

maciejbocianski commented:
#7063 main purpose is to fix `sleep_usticker_test` but it additionally turns on whole `tests-mbed_hal-sleep` test, but `deepsleep_lpticker_test` requires fix which is in #7036

`sleep_usticker_test` is disabled here, so can be merged before #7063

deepikabhavnani commented:
@maciejbocianski  @cmonr  - Added changes from https://github.com/ARMmbed/mbed-os/pull/7063 in this PR itself will save us some time in CI and sequencing.

Closing https://github.com/ARMmbed/mbed-os/pull/7063

0xc0170 commented:
This fixes the problem (the commit 0504034), not on board with the adding new hook to just overwrite the read functionality - having stats on but just return 0. Flag `MBED_CPU_STATS_ENABLED` vs this new function?
Both have the same effect.

The use for this API? One use case we have when a test is checking time requirement that stats break.

c1728p9 commented on TESTS/mbed_hal/sleep/main.cpp:
```diff
@@ -218,13 +227,19 @@ utest::v1::status_t greentea_failure_handler(const Case * const source, const fa
 #if DEVICE_LPTICKER
     lp_ticker_init();
 #endif
+    mbed_stats_ticker_read = disable_stats_read;
```
Rather than disabling the stats ticker, would it be possible the test to gracefully handle this case? This could be done either by decreasing the tolerance as in #7070 or by increasing delay time so the overhead of the stats read is proportionately less.

maciejbocianski replied:
Can be done, delta has to be increased as follows (additional 10-15us to cover worst case: nrf51):
```cpp
/* Used for regular sleep mode, a target should be awake within 10 us. Define us delta value as follows:
 * delta = default 10 us + worst ticker resolution + extra time for code execution + extra time for sleep time logging*/
static const uint32_t sleep_mode_delta_us = (10 + 4 + 5 + 10);
```

0xc0170 replied:
Looks good for now (I still would expect one day to test 10us period - no logging included - "release")

cmonr commented:
@maciejbocianski Go ahead with the tolerance change, and once this PR is updated and re-approved, we can start CI.

As a reminder, we need to get this in before https://github.com/ARMmbed/mbed-os/pull/7073 can progress, and we need it in before we can generate RC2.

maciejbocianski committed:
fix and enable sleep_usticker_test

Increases tolerance value for sleep_usticker_test to cover extra time needed
for cpu stats computation (for more details see MBED_CPU_STATS_ENABLED).
Prevent scheduling interrupt during ticker initialization (in lp_ticker_init)
while test execution.

maciejbocianski commented:
@deepika's changes has been removed
tolerance value was increased for sleep_usticker_test to cover extra time needed for cpu stats computation

cmonr commented:
/morph build

cmonr commented:
@maciejbocianski FYI. Tehe NRF52_DK and NUCLEO_F746ZG devices all failed the sleep test.

maciejbocianski commented:
My bad, not enough value of extra time for stats computation...

for NRF51_DK  all values in [us]
measured in `sleep_usticker_test`

| compiler | sleep recovery time, cpu stats off | sleep recovery time, cpu stats on |
| --------- | ------------------------------- | ------------------------------- |
| GCC  | 12 | 37/38 |
| ARM | 11 | 34/35 |
| IAR   | 12 | 37/38 |

So we have to increase extra time for stats computation to 25us to cover nrf51 worst case
and in result we get total delta 44us

```cpp
/* Used for regular sleep mode, a target should be awake within 10 us. Define us delta value as follows:
 * delta = default 10 us + worst ticker resolution + extra time for code execution + extra time for stats computation
 * (for more details see MBED_CPU_STATS_ENABLED) */
static const uint32_t sleep_mode_delta_us = (10 + 4 + 5 + 25);
```

That's what I was afraid of. We are further and further away from measured value of 10us  :worried:
No idea why `NUCLEO_F746ZG` is failing (don't have this target)

maciejbocianski commented:
following @ithinuel suggestions I changed the delta definition as follows:
```cpp
/* Used for regular sleep mode, a target should be awake within 10 us. Define us delta value as follows:
 * delta = default 10 us + worst ticker resolution + extra time for code execution */
#if defined(MBED_CPU_STATS_ENABLED)
/* extra 25us for stats computation (for more details see MBED_CPU_STATS_ENABLED) */
static const uint32_t sleep_mode_delta_us = (10 + 4 + 5 + 25);
#else
static const uint32_t sleep_mode_delta_us = (10 + 4 + 5);
#endif
```

0xc0170 commented:
/morph build

fkjagodzinski commented on TESTS/mbed_hal/sleep/main.cpp:
```diff
@@ -37,7 +37,12 @@ using namespace utest::v1;
  * delta = default 10 us + worst ticker resolution + extra time for code execution */
+#if defined(MBED_CPU_STATS_ENABLED)
+/* extra 25us for stats computation (for more details see MBED_CPU_STATS_ENABLED) */
+static const uint32_t sleep_mode_delta_us = (10 + 4 + 5 + 25);
```
:+1: for stating that clearly

maciejbocianski commented:
`NUCLEO_F746ZG` still is failing

It looks like in first attempt to sleep is failing
We go to sleep at 1633507 and expects to wake up at 1633607+delta
but we woke up at 1633520

Fix is in progress...

```
[1528194974.89][CONN][INF] found KV pair in stream: {{__testcase_name;deep-sleep - high-speed clocks are turned off}}, queued...
[1528194974.96][CONN][RXD] >>> Running case #1: 'sleep - source of wake-up - us ticker'...
[1528194975.02][CONN][INF] found KV pair in stream: {{__testcase_start;sleep - source of wake-up - us ticker}}, queued...
[1528194975.05][CONN][RXD] next_match_timestamp: 1633607
[1528194975.08][CONN][RXD] wakeup_timestamp: 1633520
[1528194975.09][CONN][RXD] diff: -87
[1528194975.10][CONN][RXD] ticks: 44
[1528194975.12][CONN][RXD] ticker_width: 32
[1528194975.17][CONN][RXD] :141::FAIL: Expression Evaluated To FALSE
[1528194975.20][CONN][RXD] next_match_timestamp: 1779539
[1528194975.23][CONN][RXD] wakeup_timestamp: 1779542
[1528194975.24][CONN][RXD] diff: 3
[1528194975.25][CONN][RXD] ticks: 44
[1528194975.27][CONN][RXD] ticker_width: 32
[1528194975.30][CONN][RXD] next_match_timestamp: 1879638
[1528194975.32][CONN][RXD] wakeup_timestamp: 1879640
[1528194975.33][CONN][RXD] diff: 2
[1528194975.34][CONN][RXD] ticks: 44
[1528194975.36][CONN][RXD] ticker_width: 32
[1528194975.40][CONN][RXD] next_match_timestamp: 1979738
[1528194975.43][CONN][RXD] wakeup_timestamp: 1979741
[1528194975.44][CONN][RXD] diff: 3
[1528194975.45][CONN][RXD] ticks: 44
[1528194975.47][CONN][RXD] ticker_width: 32
[1528194975.50][CONN][RXD] next_match_timestamp: 2079838
[1528194975.53][CONN][RXD] wakeup_timestamp: 2079841
[1528194975.53][CONN][RXD] diff: 3
[1528194975.55][CONN][RXD] ticks: 44
[1528194975.56][CONN][RXD] ticker_width: 32
[1528194975.60][CONN][RXD] next_match_timestamp: 2179938
[1528194975.62][CONN][RXD] wakeup_timestamp: 2179941
[1528194975.63][CONN][RXD] diff: 3
[1528194975.65][CONN][RXD] ticks: 44
[1528194975.67][CONN][RXD] ticker_width: 32
[1528194975.70][CONN][RXD] next_match_timestamp: 2280038
[1528194975.73][CONN][RXD] wakeup_timestamp: 2280041
[1528194975.74][CONN][RXD] diff: 3
[1528194975.75][CONN][RXD] ticks: 44
[1528194975.77][CONN][RXD] ticker_width: 32
[1528194975.80][CONN][RXD] next_match_timestamp: 2380138
[1528194975.82][CONN][RXD] wakeup_timestamp: 2380141
[1528194975.83][CONN][RXD] diff: 3
[1528194975.84][CONN][RXD] ticks: 44
[1528194975.86][CONN][RXD] ticker_width: 32
[1528194975.90][CONN][RXD] next_match_timestamp: 2480238
[1528194975.93][CONN][RXD] wakeup_timestamp: 2480241
[1528194975.94][CONN][RXD] diff: 3
[1528194975.95][CONN][RXD] ticks: 44
[1528194975.97][CONN][RXD] ticker_width: 32
[1528194976.03][CONN][INF] found KV pair in stream: {{
```

maciejbocianski commented:
`NUCLEO_F746ZG` LPTICKER is based on LPTIM and its implementation raises interrupt in `lp_ticker_init` in last line. 
https://github.com/ARMmbed/mbed-os/blob/74aec930e69d4ad1e38a138525e48b07983fad00/targets/TARGET_STM/lp_ticker.c#L156

When cpu stats are enabled, below code causes that interrupt is generated and hal_sleep returns immediately

https://github.com/ARMmbed/mbed-os/blob/74aec930e69d4ad1e38a138525e48b07983fad00/hal/mbed_sleep_manager.c#L192

0xc0170 commented:
@ARMmbed/team-st-mcd Can you review the above? Shall compare_set be removed and do in the set interrupt function ?

cmonr commented:
@SenRamakri Could you take a look at @maciejbocianski's comment about the NUCLEO_F746ZG immediate interrupt issue?

maciejbocianski commented:
fix has been already pushed

0xc0170 commented:
/morph build

OPpuolitaival commented:
build retriggered

cmonr commented:
Looks like build completed successfully, but didn't report back. Going to retrigger.

cmonr commented:
/morph build

cmonr commented:
Halting CI builds until RC3 PRs are completed. Will resume after.

0xc0170 commented:
/morph build

0xc0170 commented:
/morph build

kjbracey commented:
Looks like this has some real build failures.

maciejbocianski commented:
fix was already pushed
there was `lp_ticker` usage without `DEVICE_LPTICKER` guard

0xc0170 commented:
/morph build

cmonr commented:
Going to restart this test since the NRF51 appears to have timed out quite a bit, but there might be legitimate errors with the K82F and KW41Z building with GCC ('Low power ticker frequency test' failed for both targets).

/morph test
